keps/0029-operator-dependencies.md

```diff
@@ -0,0 +1,282 @@
+3. If we haven't found any cycles, start executing the top-level `deploy` plan. When encountering an operator-task, apply the corresponding `Instance` resource. Here it is the same as for any other resource that we create: we check if it is healthy and if not, end current plan execution and "wait" for it to become healthy. KUDO manager already has a [health check](https://github.com/kudobuilder/kudo/blob/master/pkg/engine/health/health.go#L78) for `Instance` resources implemented.
+
+Let's take a look at an example. Here is a simplified operator `AA` with a few dependencies:
+
+```text
+AA
+├── BB
+│   ├── EE
+│   │   ├── H
+│   │   └── I
+│   ├── F
+│   └── GG
+│       ├── J
+│       └── K
+├── CC
+│   ├── L
+│   └── M
+└── D
+
+Legend:
+- Operators and operator-tasks are marked with double letters e.g. 'AA' or `BB`
+- other tasks are marked with single letters e.g. 'D'
+- direct children of an operator are the 'deploy' plan steps e.g. for 'AA' deploy steps are 'BB', 'CC' and 'D'
+```
+
+In the first step we build a dependency graph. A set of all graph vertices (which are task-operators) `S` is defined as `S = {AA, BB, CC, EE, GG}`. A transitive relationship `R` between the vertices is defined as `(a, b) ∈ S` meaning _`a` needs `b` deployed first_. The transitive relationship for the above example is: `R = { (AA,BB), (AA,CC), (BB,EE), (BB,GG) }`. The resulting topological order `O` is therefor `O = (EE, GG, BB, CC, AA)` which has no cycles.
+
+The instance controller (IC) then starts with the execution of the top-level `deploy` plan of the operator `AA`. The first task is the `BB` operator-task. When executing it, IC creates the `Instance-BB` resource and ends current reconciliation. Next, IC notices new `Instance-BB` resource, starts new reconciliation, and executes the  `deploy` plan of the operator `BB` which then creates `Instance-EE` resource. This way we are basically performing the depth-first search for the dependency graph, executing each vertex in the right order e.g. `EE` has to be healthy before `BB` deploy plan can continue with the next step `F`.
+
+We would additionally add the higher-level `Instance` reference (e.g. `AA`) to the `ownerReferences` list of its direct children `Instance`s (e.g. `BB` and `CC`). This would help with determining which `Instance` belongs to which operator and additionally help us with [operator uninstalling](#uninstalling).
+
+The status of the execution can be seen as usual as part of the `Instance.Status`. We could additionally forward the status of a dependency `Instance` to the top-level `Instance.Status` to simplify the overview.
+
+Note that in the above example if e.g. `EE` and `CC` task-operators reference the same operator package they must use distinct instance names `spec.instanceName` so that two separate `Instance`s are deployed. Otherwise, the dependency graph will have a cycle.
```

Comment:
You're right, if `EE` and `CC` were the same operator it wouldn't create a cycle. But if they both reference  **the same instance** it would be a life-cycle dependency. Which would bring us in the life-cycle dependency "hell", like what if both are configured differently, or what if `EE` is part of some other operator altogether? The solution today is to create two separate instances `EE-instance` and `CC-instance`.